### DISCUSSION TITLE
Updating parameter in aging model manually after every cycle

### DISCUSSION THREAD

**Discussion by kiranbhaskar** [pybamm-team/PyBaMM#3858] [Q&A]
Hello,
I am using PyBaMM's aging models for simulating NMC battery cell aging. I am using the parameter set from SimonO'Kane et al: 

> paramA = pybamm.ParameterValues("OKane2022")

I am using the following aging mechanisms as:
`

> model = pybamm.lithium_ion.DFN(
>     {
>         "SEI": "solvent-diffusion limited",
>         "SEI porosity change": "true",
>         "lithium plating": "partially reversible",
>         "lithium plating porosity change": "true",  # alias for "SEI porosity change"
>         "particle mechanics": ("swelling and cracking", "swelling only"),
>         "SEI on cracks": "true",
>         "loss of active material": "stress-driven",
>         "calculate discharge energy": "true",  # for compatibility with older PyBaMM versions
>     }
> )`
> 

I want to modify the input current profile for every cycle depending on the instantaneous capacity of the cell after every cycle. So I am unable to use the pybamm.experiment() and have multiple cycles in that. I want to modify the cycles every iteration and it will be a function of the instantaneous capacity. So, at the end of every cycle, I am updating the following parameter (for the next cycle simulation) manually as follows:

```
# Run first cycle simulation
   sim = pybamm.Simulation(model, parameter_values=paramA, var_pts=var_pts)
    solA = sim.solve(t_eval= time)

# Update parameters:
    capacity_loss = solA["Total capacity lost to side reactions [A.h]"].entries
    Q = Q - capacity_loss[-1]

    xa = np.mean(solA["R-averaged negative particle concentration"].entries[:,-1])
    paramA['Initial concentration in negative electrode [mol.m-3]'] = np.mean(solA["R-averaged negative particle concentration [mol.m-3]"].entries[:,-1])
    paramA['Maximum concentration in negative electrode [mol.m-3]'] = paramA['Initial concentration in negative electrode [mol.m-3]']/xa
    paramA['Initial inner SEI thickness [m]'] = np.mean(solA['Inner SEI thickness [m]'].entries)
    paramA['Initial outer SEI thickness [m]'] = np.mean(solA['Outer SEI thickness [m]'].entries)
    paramA['Initial plated lithium concentration [mol.m-3]'] = np.mean(solA['Lithium plating concentration [mol.m-3]'].entries)
    paramA['Negative electrode initial crack length [m]'] = np.mean(solA['Negative particle crack length [m]'].entries)
    paramA['Positive electrode initial crack length [m]'] = np.mean(solA['Positive particle crack length [m]'].entries)
    paramA['Negative electrode porosity'] = np.mean(solA['Negative electrode porosity'].entries)
    paramA['Positive electrode porosity'] = np.mean(solA['Positive electrode porosity'].entries)

```
However, I am getting errors after a cycle: 

> File "C:\Users\Python code\venv\lib\site-packages\casadi\casadi.py", line 20021, in call
>     return _casadi.Function_call(self, *args)
> RuntimeError: Error in Function::call for 'F' [IdasInterface] at .../casadi/core/function.cpp:1401:
> Error in Function::call for 'F' [IdasInterface] at .../casadi/core/function.cpp:330:
> .../casadi/interfaces/sundials/idas_interface.cpp:596: IDACalcIC returned "IDA_FIRST_RES_FAIL". Consult IDAS documentation.
> 

Apart from the above, do you have any suggestions on how to update the initial values of the parameters for the second cycle simulation to the final values of the first cycle?

Thank you!

**Comment by martinjrobins** [comment]
Hi @kiranbhaskar. One of the steps performed by the `pybamm.Simulation` class is to replace all the parameters in the model with the values in your `paramA` object, so changing the `paramA` object after this step will have no effect on the model. If you wish to change the parameters of a model between multiple solves, the best way to achieve this is to set the parameters you wish to change to *input parameters*, you can see how this is done on this page: https://train.oxrse.uk/material/HPCu/libraries/pybamm/03_parameter_values

So you would need to change all the parameters you wish to alter to input parameters, then pass in a dictionary of the new values to every call of the solver `solve` function (https://docs.pybamm.org/en/stable/source/api/solvers/base_solver.html#pybamm.BaseSolver.solve)

**Reply by kiranbhaskar** [reply]
Thanks so much, @martinjrobins for your answer. Yes, as you mentioned, I want to simulate the aging by updating the parameters between multiple solves. I tried updating these input parameters to each solve to the output of the previous solve. But I was getting the above errors after the first solve. 
I was getting a similar error in other simulations when I had unrealistic values for my initial concentrations. But in this case, these values set to the parameters were within the allowable range as far as I know.
Do you have any suggestion on why that might be or if there is a way to debug it?
Thanks!
